service/history/nDCEventsReapplication.go

```diff
@@ -0,0 +1,262 @@
+}
+
+func (r *nDCEventsReapplication) reapplyEventsToCurrentClosedWorkflow(
+	ctx ctx.Context,
+	context workflowExecutionContext,
+	msBuilder mutableState,
+	events []*shared.HistoryEvent,
+) (retError error) {
+
+	domainID := msBuilder.GetExecutionInfo().DomainID
+	workflowID := msBuilder.GetExecutionInfo().WorkflowID
+
+	domainEntry, err := r.domainCache.GetDomainByID(domainID)
+	if err != nil {
+		return err
+	}
+
+	resetRequestID := uuid.New()
+	// workflow event buffer guarantee that the event immediately
+	// after the decision task started is decision task finished event
+	lastDecisionTaskStartEventID := msBuilder.GetPreviousStartedEventID()
+	if lastDecisionTaskStartEventID == common.EmptyEventID {
+		// TODO when https://github.com/uber/cadence/issues/2420 is finished
+		//  reset to workflow finish event
+		errStr := "cannot reapply signal due to workflow missing decision"
+		r.logger.Error(errStr)
+		return &shared.BadRequestError{Message: errStr}
+
+	}
+
+	resetDecisionFinishID := lastDecisionTaskStartEventID + 1
+
+	baseContext := context
+	baseMutableState := msBuilder
+	currContext := context
+	currMutableState := msBuilder
+	_, err = r.resetor.ResetWorkflowExecution(
+		ctx,
+		&shared.ResetWorkflowExecutionRequest{
+			Domain:                common.StringPtr(domainEntry.GetInfo().Name),
+			WorkflowExecution:     context.getExecution(),
+			Reason:                common.StringPtr(workflowResetReason),
+			DecisionFinishEventId: common.Int64Ptr(resetDecisionFinishID),
+			RequestId:             common.StringPtr(resetRequestID),
+		},
+		baseContext,
+		baseMutableState,
+		currContext,
+		currMutableState,
+	)
+	if err != nil {
+		if _, ok := err.(*shared.DomainNotActiveError); ok {
+			return nil
+		}
+		return err
+	}
+
+	resetNewContext, resetNewMsBuilder, resetNewRelease, err := r.getCurrentWorkflowMutableState(ctx, domainID, workflowID)
+	defer func() { resetNewRelease(retError) }()
+	if err != nil {
+		return err
+	}
+	if resetNewMsBuilder.IsWorkflowExecutionRunning() {
+		return ErrRetryRaceCondition
+	}
+
+	return r.reapplyEventsToCurrentRunningWorkflow(ctx, resetNewContext, resetNewMsBuilder, events)
+}
+
+func (r *nDCEventsReapplication) getCurrentWorkflowMutableState(
+	ctx ctx.Context,
+	domainID string,
+	workflowID string,
+) (workflowExecutionContext, mutableState, releaseWorkflowExecutionFunc, error) {
+	// we need to check the current workflow execution
+	context, release, err := r.historyCache.getOrCreateWorkflowExecution(ctx,
+		domainID,
+		// only use the workflow ID, to get the current running one
+		shared.WorkflowExecution{WorkflowId: common.StringPtr(workflowID)},
+	)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+
+	msBuilder, err := context.loadWorkflowExecution()
+	if err != nil {
+		// no matter what error happen, we need to retry
+		release(err)
+		return nil, nil, nil, err
+	}
+	return context, msBuilder, release, nil
+}
+
+func (r *nDCEventsReapplication) prepareWorkflowMutation(
```

Comment:
when we do reapply, we should treat the current cluster as active (close transaction function in mutable state will do domain activeness check). so this logic is not needed

this function was used in 2DC since the overall reapply logic is different

Reply:
removed this method